### ISSUE TITLE
Create an initial development version of a new TC-Diag tool to support the computation of tropical cyclone diagnostics

### ISSUE THREAD

**Issue by JohnHalleyGotway** [dtcenter/MET#2168]
## Describe the New Feature ##
The METplus TC-Diagnostics group met on May 23, 2022 to define requirements for incorporating legacy hurricane diagnostics logic in METplus. Those requirement include the creation of a new tool named TC-Diag, for Tropical Cyclone Diagnostics.

The inputs for TC-Diag are the same as those for TC-RMW:
  - a list of gridded input data files (-data)
  - an ATCF track data file (-deck)
  - a configuration file (-config)
  - an output file name (-out)

For each track point, TC-Diag transforms the gridded model data to cylindrical coordinates centered on the storm location. This can be done relative to the radius of maximum winds-space, as is done in TC-RMW. However, additional conversion logic should also be supported, such as being based on shear.

ACTION: Need to define what derivation types should be supported.

After the requested fields are read and transformed in one or more ways, provide an option to write them to an output NetCDF file. However, recommend that this option be disabled by default.

Next, incorporate python code from Robert DeMaria directly into the MET repository to actually compute the diagnostics. The TC-Diag makes calls to that python code to derive the diagnostics. The number and type of diagnostics to be computed may depend on the type of cylindrical coordinate transformation that was performed. Ideally the cylindrical coordinate data is handed to python in memory. This logic should be made highly configurable. Provide a mechanism for users to write or modify a python derivation script.

ACTION: Need to define lots of details for how MET calls the python derivation scripts and how this can be made configurable. Also need to define source of those python scripts.

The python scripts hand back to TC-Diag the computed diagnostics. These will consist of a few dozen floating point numbers for each track location.

TC-Diag writes those diagnostic values out to one of two supported formats. The desired output format must be a configurable option:

1. LS Diag (Large Scale Diagnostics) format: See sample [LS Diag formatted](https://rammb2.cira.colostate.edu/research/tropical-cyclones/ships/) files only the [NHC FTP site](https://ftp.nhc.noaa.gov/atcf/lsdiag/).
2. Model Diag format: Need to define documentation and sample data.

### Acceptance Testing ###
Input: gridded model data, ATCF track file, ASCII config file
Output: Formatted ASCII output in the LS-Diag or Model-Diag formats

### Time Estimate ###
2 weeks.

### Sub-Issues ###
Consider breaking the new feature down into sub-issues.
- [ ] *Add a checkbox for each sub-issue here.*

### Relevant Deadlines ###
The initial version of the TC-Diag tool should be ready by February 2023.

### Funding Source ###
*NOAA WPO - JTTI TCDiag Project (2770043)*

### Assignee ###
- [x] Select **engineer(s)** or **no engineer** required: @JohnHalleyGotway 
- [x] Select **scientist(s)** or **no scientist** required: @jvigh and @KathrynNewman 

### Labels ###
- [x] Select **component(s)**
- [x] Select **priority**
- [x] Select **requestor(s)**

### Projects and Milestone ###
- [x] Select **Repository** and/or **Organization** level **Project(s)** or add **alert: NEED PROJECT ASSIGNMENT** label
- [x] Select **Milestone** as the next official version or **Future Versions**

## Define Related Issue(s) ##
Consider the impact to the other METplus components.
- [x] MET: dtcenter/MET#392
- [ ] METplus: List issue here
- [ ] METdatadb: List issue here
- [ ] METviewer: List issue here

## New Feature Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [ ] Complete the issue definition above, including the **Time Estimate** and **Funding source**.
- [ ] Fork this repository or create a branch of **develop**.
Branch name: `feature_<Issue Number>_<Description>`
- [ ] Complete the development and test your changes.
- [ ] Add/update log messages for easier debugging.
- [ ] Add/update unit tests.
- [ ] Add/update documentation.
- [ ] Push local changes to GitHub.
- [ ] Submit a pull request to merge into **develop**.
Pull request: `feature <Issue Number> <Description>`
- [ ] Define the pull request metadata, as permissions allow.
Select: **Reviewer(s)** and **Linked issues**
Select: **Repository** level development cycle **Project** for the next official release
Select: **Milestone** as the next official version
- [ ] Iterate until the reviewer(s) accept and merge your changes.
- [ ] Delete your fork or branch.
- [ ] Close this issue.

**Comment by JohnHalleyGotway** [comment]
Some notes from the project discussion on 8/10/22:

- Need to create a TC-Diag tool to read GRIB data input files since WCOSS does NOT allow pygrib.
- TCDIAG must write to ASCII format while the NetCDF output is desired.
- Recommend also creating a TCDIAG line type written by TC-Pairs.
- The TC-DIAG tool.
   - Read GRIB data.
   - Calls Robert's derivation of diagnostics in Python
   - Writes ASCII output files (a NetCDF file is desired but not required).
- TC-Pairs reads them and incorporates it into a new TCDIAG output line type.

**Comment by JohnHalleyGotway** [comment]
@jvigh I've worked on the [feature_2168_tc_diag](https://github.com/dtcenter/MET/tree/feature_2168_tc_diag) branch over the last couple of days. So far all the changes have been plumbing, basically just creating a copy of the tc_rmw tool and renaming it tc_diag, adding a stub for its unit tests (unit_tc_diag.xml), and adding a largely empty chapter to the User's Guide about it.

I'm now ready to start working on substantive code changes, and I'm wondering how to get started.

I'll propose a time for us to meet to decide where to start.

**Comment by JohnHalleyGotway** [comment]
Based on our discussions, it sounds like TC-Diag will rely very heavily on python embedding in MET. As such, it should only be compiled when the `--enable-python` configuration option is chosen. Need to start figuring out the details of how we pass data back/forth between the C++ TC-Diag application and the python derivation scripts.

See this [python embedding documentation](https://docs.python.org/3/extending/embedding.html) for reference.

**Comment by JohnHalleyGotway** [comment]
Discussed at TC-Diag project meeting on 12/12/2022.
@KathrynNewman, @jvigh, and @JohnHalleyGotway 

- Should this be ALL python or a mix of C++ computation and python?
- Let's start with an ALL python approach and then evaluate whether any part of this should be done directly in C++.
- Will only be compiled when `--enable-python` is configured.
- Config file will list the python scripts to be run and the data be to read.
- Start by trying to make all the gridded data available in memory.
- Each python script would try to grab the data "by name" in memory.
- TC-Diag reads a "track" consisting of multiple points in time and process all points.
- Recommend using OPENMP to use multiple threads to process each time point independently.
- Start by working on the mechanics of reading 2 gridded fields and pass them in memory to the python scripts.
- TC-Diag should transform the data into cylindrical coordinates prior to passing to python scripts. Need some sort of data structure to store the data.
- Use gridded GFS 0.5 forecast files for development. Use and Ian track for development (20220923_06 through 20220930_12, every 6 hours). Pull 0.5 degree GFS forecasts from 20220923_06 20220930_12.
- Need to replicate the existing ASCII diagnostic file format.
- Define the grid using these config options:
```
//
// Range-Azimuth grid parameters
//
n_range        = 100;
n_azimuth      = 180;
max_range_km   = 1000.0;
delta_range_km = 10.0;
rmw_scale = 0.2;
```
- Recommend allowing the cylindrical coordinates to be defined separately for each field being read.
- Recommend doing a careful comparison between the existing python cylindrical coordinate transformation and comparing that to MET's output.
- Have TC-Diag dump it's transformed gridded data to a NetCDF file as an option. Model this after the existing TC-RMW output. Factor out common functionality where possible into library code.

Task checklist:
- [x] List `convert(x)` in the TC-Diag config file since it already works anyway.
- [x] Make compilation of TC-Diag dependent on `--enable-python`. Updates in `src/tools/tc_utils/Makefile.am`
- [x] Update the config options to include TCDiagNcOutInfo, TCDiagDataOpt, and TCDiagConfInfo. Define TCDiagDataOpt for each data field so that we can eventually support multiple cyclindrical coordinate definitions in a single run.
- [x] Apply consistent whitespace formatting and confirm that the feature branch is compiling again on 12/21/22.
- [ ] Rather than supporting an arbitrary number of range/azimuth grids specified separately for each input field, support only 2 options. Transform that data using fixed range values or letting them vary relative to RMW.
- [ ] ADD TASKS to the list as they're completed here.

**Comment by jvigh** [comment]
Let's preserve this chat discussion in the issue and keep the thread going here. 
![image](https://user-images.githubusercontent.com/19677014/209213279-5908f45e-5a34-4956-bc5b-b20aa6e3d660.png)

**Comment by JohnHalleyGotway** [comment]
During the TC-Diag project meeting on 2/13/2023, we did some additional brainstorming about the desired functionality, as described in [these notes](https://docs.google.com/document/d/1ZO3pPHIe6FAW6mUZWBA2wTcI3jF6pzT6VtvUed6sgyo/edit#heading=h.e8dr2x6blirh).

Here are some important details:
- TC-Diag will behave a lot like TC-RMW. For each track point lead time, read all the gridded data requested in the configuration file into memory. Perform the desired transformation to cylindrical coordinates and write the output to a temporary NetCDF file FOR THAT TIME.
- Using python embedding, execute each of the python diagnostic definition scripts listed in the configuration file. To simplify things, we might as well make the location of the temporary file to be read an argument to each of the scripts.
- Consider HOW each of the scripts should read data from that temporary NetCDF file. Consider writing a common python module to do so that lives either in the MET repository itself or in METcalcpy... probably MET itself until/unless there's a clear need to move it to METcalcpy.
- Also note that the "standard" python scripts for generating diagnostics should live in MET itself.
- Each python script should store the derived diagnostics in some standard object name (just like all the other python embedding examples do). And MET just extracts the values from the object name.
- The diagnostics for each of the track point lead times are INDEPENDENT. Recommend running them in parallel. Would openmp suffice or would we need to introduce threads?
- Likely need a diagnostic definition script type that processes only the track file rather than the NetCDF file.
- Provide options to write output in ascii (tc diagnostics format and/or LS_DIAG?) and NetCDF.
- Provide option to COMBINE those temporary NetCDF files into a single output one prior to deleting them.
- Note that MET currently reads data from a gridded input file separately for each VarInfo object. This is probably too inefficient for this use since that'll require us to open/inventory/close that file FOR EACH field array entry. Recommend that we create an alternative to read ALL THE DATA for ALL VAR INFO objects from the file at once. We can store the result in a vector of DataPlaneArray objects.

**Comment by JohnHalleyGotway** [comment]
As discussed on 2/27/23, recommend that a single call to TC-Diag be able to process multiple tracks. This would be most useful, for example, when processing multiple ensemble members for the same initialization time.

Enable TC-Diag to process multiple tracks, but only for a single initialization time.

**Comment by JohnHalleyGotway** [comment]
Based on the discussion on 3/6/23, consider the following:

1. Per Kate, the most common workflow is to generate diagnostics for all "current" storms.
  - Consider letting tc_diag read in "all" the model data for a single initialization time.
  - But let it process tracks for multiple storms and/or multiple models for that same initialization time.
  - Should I let the data to be read vary for each domain. Perhaps the small number of diagnostics computed for subdomains require a smaller set of data fields?
2. The temp files generated should include the storm id, track, and the grid.
3. There can be an arbitrary number of nests. Currently, the majority of the diagnostics are computed over the parent domain. Only 2 of them are computed from the nest. Currently only 1 parent and 1 nest are created, but in the future, there will be an arbitrary number of nests, one for each active storm.
  - Recommend updating the `-data` command line option to allow for multiple data sources to be provided. But we do need to differentiate between them by name, i.e. "parent" vs "nest"... or "global" vs "al092022".
4. Support the option of writing the model diagnostics to a "small" NetCDF file.
5. Support the option of writing the model diagnostics AND the transformed gridded model data to a "large" NetCDF file.

**Comment by JohnHalleyGotway** [comment]
@jvigh and @KathrynNewman, I'm porting over our conversation from Google Chat to this issue.

I know we talked about tc-diag output files some last week. For each track processed, we want the ability to write:
1. an ascii CIRA diagnostics output file
2. a NetCDF diagnostics output file
3. a NetCDF file containing gridded cylindrical coordinate data

I'm wondering if 2. and 3. should be combined into a single NetCDF file? Is that simpler? Or is it simpler for the user to have the cylindrical coordinate data used to compute the diagnostics in a DIFFERENT file from the actual diagnostics that were computed? I'm thinking in practice, users may want to write both 2. and 3., but throw the larger 2. data files away while keeping the smaller 3. diagnostic data files for a longer time period.

---

@jvigh confirms that keeping them separate is good. That provides maximum flexibility at the expense of a few inodes on disk (for numerous small files). But as long as we have one entire initialization's worth of stuff in a file, I think that's fine. I'd be against writing out each separate time to a separate file as that's going to quickly build up to the millions of files each season.

---

In order to parallelize the processing as much as possible we will want to write a separate temporary NetCDF file with the cylindrical coordinates data for each individual track point. And that would translate into the "millions" of files for each season. I'll plan to update TC-Diag so that... 
- If the user does NOT want the cylindrical coordinate data, TC-Diag will just delete those temp files.
- If the user DOES want the cylindrical coordinate data, it'll combine those TrackPoint-specific temp files into a single NetCDF file for each track prior to deleting the temp files.

**Comment by JohnHalleyGotway** [comment]
@jvigh and @KathrynNewman, I have a TC-Diag logic question for you. It is USUALLY the case that model output is stored in separate files for each valid time. For example, GFS output is typically organized into 1 file for the analysis time, 1 file for the 6-hour forecast, 1 file for the 12-hour forecast, and so on.

However, I have run across files that contain multiple output times. For example, output for all forecast hours are stored in a single file.

Should I structure the logic for TC-Diag to optimize for the former but allow for the latter? Or should I not worry about the latter at this point and just make the former work?

**Comment by JohnHalleyGotway** [comment]
From Robert DeMaria on April 4, 2023 to define the specifics:

1) What parameters define the nested cylindrical grid?
150 radii with a spacing of 2km where the azimuth is divided into 8 equally spaced segments. (This is the same as the outer grid except that the radii spacing is 2km instead of 10km).

2)What are the grib variables needed for the diagnostic calculations?
u, v, t, r, gh with level type isobaricInhPa at each pressure level
2t, 2r with level type heightAboveGround at level 2
prmsl with level type meanSea at level 0
10u, 10v with level type heightAboveGround at level 10
pwat with level type unknown 
t with level type surface at level 0

ATCF variables: lon, lat, vmax, mslp

3) What are the values used to define the azimuths of the cylindrical grid?
In degrees:
  0.,  45.,  90., 135., 180., 225., 270., 315.

4) Should fraction of RMW ever be used to define the radii of the cylindrical grid?
No, this should be removed for the time being.  If demand exists for this, we can alter our code to support this later.

**Comment by JohnHalleyGotway** [comment]
From Kate Musgrave on 5/18/23:

Just to have an email record - the default "mandatory levels" for diagnostic calculations (not including the surface) are
https://www.lakeeriewx.com/Reference/MandatoryPressureLevels.html
plus 100 mb.

So:
1000, 925, 850, 700, 500, 400, 300, 250, 200, 150, 100 mb

While this is the default, users can specify any list whether or not it includes these.  There are a few calculations (shear, etc.) that rely on specific levels so if those levels are missing the dependent calculations will be too.  That logic may not be built in yet, as I don't believe we anticipated people not having the minimum levels - I'll check with Robert.

**Comment by JohnHalleyGotway** [comment]
At the METplus Engineering on May 30, 2023, we discussed that we would like to include an initial version of the tc_diag tool in MET version 11.1.0, but temper the expectations of users as to its maturity. Recommend including a disclaimer about tc_diag to indicate that it is **beta**, meaning that it is still under development and refinement. The beta label will be removed in a future release, likely version 12.0.0.

Recommend writing a common library function to print the beta **WARNING** message so that it can easily be re-used in the future. Print that warning every time tc_diag is run, including when printing the usage statement. Also include it in the documentation.

@jvigh please review this proposal and comment here if you disagree.
